### PR TITLE
Optimize XxHash3

### PR DESCRIPTION
**Pull request by xoofx**
Hey there!
This is a PR to optimize XxHash3 for large buffers (> 240 bytes) by trying to avoid loading/storing the accumulators on each 64 bytes but instead doing it after a whole batch.

```
BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22000.1098/21H2)
AMD Ryzen 9 5950X, 1 CPU, 32 logical and 16 physical cores
.NET SDK=7.0.100-rc.1.22431.12
  [Host]     : .NET 7.0.0 (7.0.22.42610), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.0 (7.0.22.42610), X64 RyuJIT AVX2


|        Method |          data |     Mean |    Error |   StdDev | Ratio |
|-------------- |-------------- |---------:|---------:|---------:|------:|
|    XXH3Native | Byte[1048576] | 27.81 us | 0.070 us | 0.062 us |  0.43 |
|          XXH3 | Byte[1048576] | 64.14 us | 0.102 us | 0.085 us |  1.00 |
| XXH3Optimized | Byte[1048576] | 31.89 us | 0.070 us | 0.059 us |  0.50 |
```

- `XXH3Native` is the C++ version 
- `XXXH3` is the current C# version
- `XXH3Optimized` is this PR C# version

So overall, this PR brings x2 performance improvement on large buffers and lower the performance gap with the native version (~10%).

Small note not covered by this PR: while checking with the native version, I noticed that the hash is different (for both the current and optimized version vs native). Will have to dig further why.

### CODE REVIEW

**Comment by xoofx** [comment]
Another note: When I checked the implementation on a macOS/m1Pro I noticed that the ARM64 implementation is significantly slower than the x86 version (e.g x3). I haven't checked, but I saw on the xxHash repository that they are doing things slightly differently for ARM platform (e.g maybe using aligned loads). Will have to dig this one as well.

**Comment by EgorBo** [comment]
> I noticed that the ARM64 implementation is significantly slower than the x86 version

I bet it's because of this: https://github.com/dotnet/runtime/pull/76641#discussion_r987378273

`operator *` for `Vector128<ulong>` ends up in a slow fallback because it doesn't exist on arm cc @stephentoub

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs [hunk 1]
```diff
@@ -572,13 +572,8 @@ private static ulong HashLengthOver240(byte* source, uint length, ulong seed)
...
 
-                uint offset = 0;
-                for (int i = 0; i < blocksNum; i++)
-                {
-                    Accumulate(accumulators, source + offset, secret, StripesPerBlock);
-                    ScrambleAccumulators(accumulators, secret + (SecretLengthBytes - StripeLengthBytes));
-                    offset += BlockLen;
-                }
+                Accumulate(accumulators, source, secret, StripesPerBlock, true, blocksNum);
+                var offset = BlockLen * blocksNum;
```
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs:576]: var => actual type

... here and all the other places...

**Comment by stephentoub** [comment]
> I noticed that the hash is different (for both the current and optimized version vs native). Will have to dig further why.

What native implementation are you calling and what are you comparing it against?  I'd validated all of the test data against a locally built xxhash.dll calling XXH3_64bits_withSeed.

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs [hunk 2]
```diff
@@ -892,4 +1009,4 @@ private struct State
             internal ulong Seed;
         };
     }
-}
+}
```
Comment (stephentoub) [inline src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs:1012]:
```suggestion
}

```

**Comment by stephentoub** [comment]
> > I noticed that the ARM64 implementation is significantly slower than the x86 version
> 
> I bet it's because of this: [#76641 (comment)](https://github.com/dotnet/runtime/pull/76641#discussion_r987378273)
> 
> `operator *` for `Vector128<ulong>` ends up in a slow fallback because it doesn't exist on arm cc @stephentoub Namely, this line: https://github.com/stephentoub/runtime/blob/c3056f711fe315e71dc6e90ccd4184674bb01a73/src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs#L784

Yeah, you can see the comment I left in the code about it and my question about better ideas for dealing with it:
https://github.com/dotnet/runtime/pull/76641#discussion_r987342213

File: src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs (hunk 1 above)
Comment (xoofx) [reply src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs]: Fixed by 2b32e8a

**Comment by xoofx** [comment]
> What native implementation are you calling and what are you comparing it against? I'd validated all of the test data against a locally built xxhash.dll calling XXH3_64bits_withSeed.

Yes, I validated against a locally built xxhash.
Now, I was just casting to a `uint` but generating the full hash gives me the following:

```
Native: 0xd36c0e13a3df139e
C#:     0x9e13dfa3130e6cd3
```

The C# bytes are read from the destination span, casting to a `Span<ulong>`. The native version is what is returned by the API.

So the bytes are actually in reverse order. Is it expected?

**Comment by xoofx** [comment]
Tests are failing, will dig further into these errors tonight.

**Comment by stephentoub** [comment]
> So the bytes are actually in reverse order. Is it expected?

Yes.
https://github.com/dotnet/runtime/blob/9a138bf5cb4a9055e4e0370a3eb3883514555fcd/src/libraries/System.IO.Hashing/src/System/IO/Hashing/XxHash3.cs#L21
https://github.com/dotnet/runtime/issues/76279

FYI, @bartonjs. I'm not alone here ;-)

**Comment by xoofx** [comment]
> Tests are failing, will dig further into these errors tonight.

Commit b72972c should fix the tests failing on the scalar path.

**Comment by xoofx** [comment]
> FYI, @bartonjs. I'm not alone here ;-)

Not sure I understand the discussion about the endianess (intermixed with whether it should be commented on the XML doc API) but I concur that using Big Endian is really confusing and counter intuitive. Who is going to expect this frankly? Last time I programmed a big endian processor that was 30 years ago with the Motorola 68000 😅 

XXHash is following the native endian order (except for 128 bits where they are using 2 64 bits) and I would hope that we could keep the same logic so that is doesn't cause such trouble when trying to match with the C++ implementation.

**Comment by stephentoub** [comment]
> XXHash is following the native endian order

I don't see any APIs that are outputting bytes, only numbers.  Can you elaborate?

**Comment by xoofx** [comment]
> I don't see any APIs that are outputting bytes, only numbers. Can you elaborate?

Sorry, I meant that they output numbers (except for the 128 bits which are 2x64 bit numbers) that thus follow the native endianness. Unless there is something in the .NET HashAlgorithm that mandates that all of them should always emit byte[] in big endian form?

**Comment by xoofx** [comment]
> "For systems which require to store and/or display the result in binary or hexadecimal format, the canonical format is defined to reproduce the same value as the natural decimal format, hence follows big-endian convention (most significant byte first)."

Oh, right, I can see it mentioned in the [header file here](https://github.com/Cyan4973/xxHash/blob/2de0fd6d9b3fd8bbb35ebd52152ecd11fc1ebbfc/xxhash.h#L661-L680)

That makes sense then, thanks!

**Comment by stephentoub** [comment]
FWIW, I don't see 50% improvement.  On my machine the best I see is ~25%:

| Method |         Toolchain |    Size |         Mean | Ratio |
|------- |------------------ |-------- |-------------:|------:|
|   Hash | \main\corerun.exe |     240 |     20.05 ns |  1.00 |
|   Hash |   \pr\corerun.exe |     240 |     20.10 ns |  1.00 |
|        |                   |         |              |       |
|   Hash | \main\corerun.exe |     512 |     31.55 ns |  1.00 |
|   Hash |   \pr\corerun.exe |     512 |     27.77 ns |  0.88 |
|        |                   |         |              |       |
|   Hash | \main\corerun.exe |    1024 |     52.49 ns |  1.00 |
|   Hash |   \pr\corerun.exe |    1024 |     41.17 ns |  0.78 |
|        |                   |         |              |       |
|   Hash | \main\corerun.exe | 1048576 | 83,650.94 ns |  1.00 |
|   Hash |   \pr\corerun.exe | 1048576 | 61,600.73 ns |  0.74 |

Still, a nice bump.  I don't love the manual loop unrolling and associated code increase, but it's probably still worth it.

Thanks.

**Review by stephentoub — APPROVED** [review]

**Comment by xoofx** [comment]
> FWIW, I don't see 50% improvement. On my machine the best I see is ~25%:

Oh interesting. What's your CPU?

**Comment by stephentoub** [comment]
```
BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22621.674)
Intel Core i7-8700 CPU 3.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
```

**Comment by xoofx** [comment]
> ```
> BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22621.674)
> Intel Core i7-8700 CPU 3.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
> ``` 

Thanks, I don't have a recent Intel CPU, but I will check on a older Kaby-Lake how it behaves.

In the meantime, I received a Windows Dev Kit 2023 with ARM64 and the results are:

```
BenchmarkDotNet=v0.13.2, OS=Windows 11 (10.0.22621.755)
Snapdragon Compute Platform, 1 CPU, 8 logical and 8 physical cores
.NET SDK=7.0.100-rc.2.22477.23
  [Host]     : .NET 7.0.0 (7.0.22.47203), Arm64 RyuJIT AdvSIMD
  DefaultJob : .NET 7.0.0 (7.0.22.47203), Arm64 RyuJIT AdvSIMD


|        Method |          data |      Mean |    Error |   StdDev | Ratio |
|-------------- |-------------- |----------:|---------:|---------:|------:|
|    XXH3Native | Byte[1048576] |  70.16 us | 0.092 us | 0.086 us |  0.25 |
|          XXH3 | Byte[1048576] | 284.62 us | 0.882 us | 0.782 us |  1.00 |
| XXH3Optimized | Byte[1048576] | 156.05 us | 0.089 us | 0.074 us |  0.55 |
```

Which gives a similar performance boost to x2 times, but it is still 2x slower than the native version.

**Pull request merged**